### ISSUE TITLE
`democracy.delegations.entries()` doesn't work at kusama height 295787

### ISSUE THREAD

**Issue by wliyongfeng** [polkadot-js/api#4815]
Network: kusama.
Tried api version: 8.0.2, 8.1.2

My code to test:
```javascript
const { getApi } = require("../api");
const { encodeAddress } = require("@polkadot/util-crypto");

;(async () => {
  const api = await getApi();
  const height = 295787;
  const blockHash = await api.rpc.chain.getBlockHash(height);
  const blockApi = await api.at(blockHash);

  const delegations = await blockApi.query.democracy.delegations.entries(); // the code which produce error
  console.log(delegations);
})()

```

The error produced:

> Error: Unable to decode storage democracy.delegations: entry 0:: createType((AccountId,Conviction)):: (AccountId,Conviction):: Decoded input doesn't match input, received 0xca5bc1915da74aba3aadd7ce7b809045d5eb5b73559259755fdcd85a40a5dc6e060000 (35 bytes), created 0xca5bc1915da74aba3aadd7ce7b809045d5eb5b73559259755fdcd85a40a5dc6e06 (33 bytes)

Same bug also happens at heights 433879, 452692, 504000, 504329, etc. What I'm trying to do is calculating the tally for kusama historical referendums while you know there is no direct storage for these data.

**Comment by jacogr** [comment]
Weird will do some digging, not sure what the extra byte would be, but certainly a type mismatch on these.

**Comment by jacogr** [comment]
This is quite interesting, checked all the blocks you provided the last byte always maps to the `Locked6x` entry, i.e. it decoded `06` and then there are 2 additional bytes following. Also checked against the current on-chain types, `Locked6x` certainly doesn't have an additional `u16` (2 bytes following). 

Will have to go back in the Rust code to see what is up here, the Enum certainly looks correct based on my recollection and based on the _current_ on-chain types, but something is not right, i.e. conviction should be a single byte in all cases, not 3. (The first 32 bytes is the accountId, so all good there)

**Comment by wliyongfeng** [comment]
Over my head, thanks.

**Comment by jacogr** [comment]
Ok, figured out what the issue is - the entry is actually an old-style linked-map, not a normal map. So either the `.entries()` doesn't handle it correctly or it should not actually be available on it. 

(I will also need to try and remember how to traverse these in the JS API, it has been ages since I used them)

**Comment by wliyongfeng** [comment]
> So either the .entries() doesn't handle it correctly or it should not actually be available on it.

It means linked-map storage should not have `.entries()` interface. Is my understanding correct?

> (I will also need to try and remember how to traverse these in the JS API, it has been ages since I used them)

Please tell me when you recall it, 🙏.

**Comment by jacogr** [comment]
Digging :)

**Comment by jacogr** [comment]
Ok... so, with this issue, this PR was a very bad idea... https://github.com/polkadot-js/api/pull/2643/files - it effectively removed support completely, since well, all chains have dropped use. However, obviously when doing that I didn't think of historic queries.

So, will have to bring back support for it. In this case however, will try my best to make it operate in the same `.entries()` & `.keys()` fashion so it is just transparent to the user. Do bear in mind that they are slower, you basically get them in sequence since each key has an entry to the next. (Which is why support was dropped on the Rust side, they are very ineffective)

TL;DR Have the issue, will get something in place to make this possible.

**Comment by wliyongfeng** [comment]
> Do bear in mind that they are slower, you basically get them in sequence since each key has an entry to the next. (Which is why support was dropped on the Rust side, they are very ineffective)

Understand for a linked-map data structure, but I think there are no many delegations data in the early blocks from my debugging, so I guess it will be fine just for this storage.

**Comment by jacogr** [comment]
Please do check against the current API version - it now correctly returns the `Linkage` results for older blocks.

**Comment by wliyongfeng** [comment]
Confirmed, I can decode the old delegations now. Thanks!

**Comment by wliyongfeng** [comment]
Sorry, I got another question.  Do you know how to decode the address from the `linked_map` storage key? I got a 64 byte key:
`0xf2794c22e353e9a839f12faab03a911bf6eed5c65f50198f0f53f8f499f770913238ba72fc05ae6b1375289fc3333314599f674fbc27678f30179fcbdaf4bc3b`

I use following code to decode the delegator(key) address(72byte).
```javascript
const delegator = encodeAddress(storageKey.slice(40), 2)
```

But it can't produce correct delegator address for legacy storage key(64byte).
```javascript
const delegator = encodeAddress(storageKey.slice(32), 2)
```


FYI: The corresponding rust code

old code(64 byte key):
```rust
/// Get the account (and lock periods) to which another account is delegating vote.
pub Delegations get(fn delegations): linked_map T::AccountId => (T::AccountId, Conviction);
```

relatively new code(72 byte key):
```rust
/// Get the account (and lock periods) to which another account is delegating vote.
pub Delegations get(fn delegations):
     linked_map hasher(blake2_256) T::AccountId => (T::AccountId, Conviction);
```

**Comment by jacogr** [comment]
The old keys are non-decodable. They used the blake2 hash, no concat - I saw that in my tests and confirmed with the old Rust code. 

`.entries()` actually returns the keys in decoded form, you can inspect the args - obviously only decodable when a decodable hasher is used.

**Comment by wliyongfeng** [comment]
I see. A little curious how the substrate storage works while the rust code can just enumerate with the `delegator` key though.

```javascript
<Delegations<T>>::enumerate()
	.filter(|(delegator, (delegate, _))|
		*delegate == to && !<VoteOf<T>>::contains_key(&(ref_index, delegator.clone()))
	)
```
But it's fine enough now, I can solve my problem to check the known delegate extrinsic(only one).
